linkerd/proxy/balance/src/discover.rs

```diff
@@ -1,31 +1,85 @@
-use futures::prelude::*;
-use linkerd_proxy_core::Resolve;
+use futures::{ready, Stream};
+use linkerd_error::Error;
 use linkerd_stack::NewService;
-use std::net::SocketAddr;
+use pin_project::pin_project;
+use std::{
+    fmt::Debug,
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::discover::Change;
+use tracing::{debug, info};
 
-pub mod buffer;
-pub mod from_resolve;
-pub mod new;
+pub(super) mod buffer;
+mod from_resolve;
 
-pub use self::{from_resolve::FromResolve, new::DiscoverNew};
+pub(super) use self::from_resolve::FromResolve;
 
-pub type Buffer<S> = buffer::Buffer<SocketAddr, S>;
+// Prepares a discovery update for the balancer by turning it into a Service.
+#[pin_project]
+pub struct NewServices<T, N> {
+    buffer: buffer::Buffer<SocketAddr, T>,
+    new: N,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Discovery stream lost")]
+pub struct DiscoveryStreamOverflow(());
+
+// === impl NewServices ===
+
+impl<T, N> NewServices<T, N> {
+    pub(crate) fn new(buffer: buffer::Buffer<SocketAddr, T>, new: N) -> Self {
+        Self { buffer, new }
+    }
+}
 
-pub(crate) fn spawn_new_from_resolve<T, R, M, N>(
-    capacity: usize,
-    resolve: R,
-    new_service: M,
-    target: T,
-) -> Buffer<N::Service>
+impl<T, N> Stream for NewServices<T, N>
 where
-    T: Clone,
-    R: Resolve<T>,
-    M: NewService<T, Service = N>,
-    N: NewService<(SocketAddr, R::Endpoint)> + Send + 'static,
-    N::Service: Send,
+    T: Clone + Debug,
+    N: NewService<(SocketAddr, T)>,
 {
-    let new_endpoint = new_service.new_service(target.clone());
-    let resolution = resolve.resolve(target).try_flatten_stream();
-    let disco = DiscoverNew::new(FromResolve::new(resolution), new_endpoint);
-    buffer::spawn(capacity, disco)
+    type Item = Result<Change<SocketAddr, N::Service>, Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Change<SocketAddr, N::Service>, Error>>> {
+        let this = self.as_mut().project();
+
+        // If the buffer has lost the ability to process new discovery updates
+        // from its resolution, blow up the entire balancer without processing
+        // further updates from the channel.
+        if this
+            .buffer
+            .overflow
+            .load(std::sync::atomic::Ordering::Acquire)
+        {
+            return Poll::Ready(Some(Err(DiscoveryStreamOverflow(()).into())));
+        };
+
+        // Process any buffered updates.
+        let change_tgt = match ready!(this.buffer.rx.poll_recv(cx)) {
+            Some(Ok(c)) => c,
+            Some(Err(e)) => return Poll::Ready(Some(Err(e))),
+            None => return Poll::Ready(Some(Err(DiscoveryStreamOverflow(()).into()))),
```

Comment:
we drop the `mpsc::Sender` in two cases: one, when we observe an actual error, which we send before we drop the sender, and two, when the stream overflows. so, in the error case, we'll receive the error first here and return...but the subsequent `poll_next` call will return the `DiscoveryStreamOverflow` error rather than `None` to indicate that the stream terminated due to an unrecoverable error. 

i don't see any obvious issues that result from that behavior, but wanted to point it out. we could, potentially, add a boolean that tracks whether we've seen an error from the stream or not, and only return `DiscoveryStreamOverflow` on `None`s if we haven't. or, we could change the `None` arm to check the `buffer.overflow` bool and only return the overflow error if it's been set...i'm not sure if this is actually necessary, but i wanted to draw attention to it.

Reply:
> in the error case, we'll receive the error first here and return...but the subsequent poll_next call

If the stream returns an error there will not be a subsequent call. Errors are considered fatal.